### PR TITLE
Accept iterator in device dmatrix.

### PR DESCRIPTION
**Pull request by trivialfis**
Stacked on #5782 and #5781 .  Replacing #5629 .

Related https://github.com/dmlc/xgboost/issues/5143 .

Related #5583 .

* Accept iterator for device dmatrix.
A new iterative device dmatrix is implemented, replacing the old device dmatrix.
* No longer have test for column major, as we now treat them uniformly as dense data.
* Add a custom data iterator for handling single batch of data.  
* Now the device quantile dmatrix can do weighted sketching.

### CODE REVIEW

**Comment by trivialfis** [comment]
> can we deprecate old APIs and move to new unified APIs

No. That will require a CPU implementation.  It's possible with the basic structure defined in this PR, but I won't carry it out, at least not in near future.


> What is the relationship between adapters and this new iterator API? Are adapters still needed and in what circumstances?

Iterator returns batches of data, adapter adapts each batch.

File: src/data/array_interface.h [hunk 1]
```diff
@@ -230,6 +230,7 @@ class ArrayInterfaceHandler {
 
 // A view over __array_interface__
 class ArrayInterface {
+ private:
```
Comment (RAMitchell) [inline src/data/array_interface.h:233]: This looks redundant.

File: src/common/quantile.h [hunk 2]
```diff
@@ -475,30 +475,6 @@ class QuantileSketchTemplate {
         this->data = dmlc::BeginPtr(space);
       }
     }
-    /*!
```
Comment (RAMitchell) [inline src/common/quantile.h:478]: Not sure if this change is relevant.

Comment (trivialfis) [reply src/common/quantile.h]: Reverted

File: src/data/array_interface.h (hunk 1 above)
Comment (trivialfis) [reply src/data/array_interface.h]: Reverted

**Pull request merged**